### ISSUE TITLE
Null gradient when turning free_graph off

### ISSUE THREAD

**Issue by eliemichel** [mitsuba-renderer/enoki#98]
The following snippet prints null gradients while if using `backward(c, true)` we get the right value (5.0, 2.0):

```C++
using FloatD = DiffArray<float>;
FloatD a = 2.0f;
FloatD b = 5.0f;
set_requires_gradient(a);
set_requires_gradient(b);

FloatD c = a * b;

backward(c, false);
LOG << "dc/da = " << gradient(a);
LOG << "dc/db = " << gradient(b);
```

Built with MSVC16, without CUDA, commit e240a4b

**Comment by Speierers** [comment]
Hi @eliemichel ,

I doubt `DiffArray<float>`. IIRC automatic differentiation in enoki is only supported for `CUDAArray`. Do you have the same issue when using `DiffArray<CUDAArray<float>>` instead?

**Comment by eliemichel** [comment]
Regarding the other issue I don't have precise ideas but for this one what do you think about the suggested fix of changing line 896 of autodiff.cpp? Did I misunderstand the meaning of this `ref_count_int`?
